### PR TITLE
feature/patro23-7 sign in

### PR DESCRIPTION
**Pull request by LukaszWinowicz**
https://tracker.intive.com/jira/browse/PATRO23-7

### CODE REVIEW

File: src/api/app/appsettings.Development.json
Comment (pzajaczkowski) [inline src/api/app/appsettings.Development.json:1]: Changes unnecessary after merge with dev (except of credentials if we want to change client authentication in keycloak)

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]
After merge with newest dev i managed to test sign in and i can say it works. I recommend to do the same before merge, and decide if we want to use public client or authenticated (with secret). In case of second option you would have to change inbudget-client settings, and then copy client secret, export new realm settings and paste secret inside. Approach with public client works either btw.

File: src/api/app/Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	/// An HTTP response containing a JWT token if the sign-in was successful, or an error response
+	/// if the sign-in failed. <see cref="Task"/>Representing the asynchronous operation.
+	/// </returns>
+	/// <exception cref="AppException">
+	/// Thrown if one or more errors occur while trying to authenticate the user.
+	/// </exception>
+	/// <response code="200">Successfully signed in.</response>
+	/// <response code="401">Username or password is not valid.</response>
+	/// <response code="500">Internal server error.</response>
+	[ProducesResponseType(typeof(string), StatusCodes.Status200OK)]
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/Controllers/UserController.cs:47]: I changed according to the comment.

File: src/api/app/Controllers/UserController.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		{
+			HttpResponseMessage response = await this.queryBus.Query<SignInUser, HttpResponseMessage>(command);
+
+			if (response.StatusCode == HttpStatusCode.Unauthorized)
+			{
+				return this.Unauthorized();
+			}
+
+			string responseContent = await response.Content.ReadAsStringAsync();
+			if (responseContent != null)
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/Controllers/UserController.cs:65]: Done.

File: src/api/app/User/Queries/SignInUser.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
+using Intive.Patronage2023.Api.Configuration;
+using Intive.Patronage2023.Shared.Abstractions.Queries;
+using Microsoft.Extensions.Options;
+
+namespace Intive.Patronage2023.Api;
+
+/// <summary>
+/// SignIn user query.
+/// </summary>
+/// <param name="Username">Username which user provides.</param>
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/User/Queries/SignInUser.cs:10]: I changed in the whole code from username to email.

File: src/api/app/Controllers/UserController.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		this.queryBus = queryBus;
+	}
+
+	/// <summary>
+	/// Authenticates a user and returns a JWT token.
+	/// </summary>
+	/// <param name="command">The sign-in command, which includes the username and password.</param>
+	/// <returns>
+	/// An HTTP response containing a JWT token if the sign-in was successful, or an error response
+	/// if the sign-in failed. <see cref="Task"/>Representing the asynchronous operation.
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/Controllers/UserController.cs:39]: I removed.

File: tests/unit/Intive.Patronage2023.Example.Domain.Tests/Intive.Patronage2023.Example.Domain.Tests.csproj
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply tests/unit/Intive.Patronage2023.Example.Domain.Tests/Intive.Patronage2023.Example.Domain.Tests.csproj:1]: Fixed.

File: src/api/app/Controllers/UserController.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
+using System.Net;
+using FluentValidation;
+using Intive.Patronage2023.Api.User;
+using Intive.Patronage2023.Modules.Example.Application.Example;
+using Intive.Patronage2023.Shared.Abstractions.Queries;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Newtonsoft.Json;
+
+namespace Intive.Patronage2023.Api;
```
Comment (DawidBski) [inline src/api/app/Controllers/UserController.cs:10]: change namespace

File: src/api/app/User/Queries/SignInUserValidator.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+public class SignInUserValidator : AbstractValidator<SignInUser>
+{
+	/// <summary>
+	/// Initializes a new instance of the <see cref="SignInUserValidator"/> class.
+	/// </summary>
+	public SignInUserValidator()
+	{
+		this.RuleFor(signin => signin.Email).NotEmpty().NotNull();
+		this.RuleFor(signin => signin.Password).NotEmpty().NotNull();
+	}
```
Comment (DawidBski) [inline src/api/app/User/Queries/SignInUserValidator.cs:17]: The email address and password must meet more requirements according to the current Keycloak configuration.

Comment (LukaszWinowicz) [reply src/api/app/User/Queries/SignInUserValidator.cs:17]: We don't need that validation for sign in, we only need it when sign up an account.

File: src/api/app/appsettings.Development.json
Comment (pzajaczkowski) [inline src/api/app/appsettings.Development.json:1]: Merge dev before further testing. All changes in this file from remote

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]

File: src/api/app/appsettings.Development.json
Comment (LukaszWinowicz) [reply src/api/app/appsettings.Development.json:1]: After talk with @mateuszmuszynski we don't want credentials. I removed them.

**Review by pzajaczkowski — APPROVED** [review]

File: tests/unit/Intive.Patronage2023.Example.Domain.Tests/Intive.Patronage2023.Example.Domain.Tests.csproj [hunk 7]
```diff
@@ -1,34 +1,34 @@
 <Project Sdk="Microsoft.NET.Sdk">
-<Import Project="..\..\..\app.props" />
```
Comment (mateuszmuszynski) [inline tests/unit/Intive.Patronage2023.Example.Domain.Tests/Intive.Patronage2023.Example.Domain.Tests.csproj:2]: whitespace changes, please revert it. We wouldn't like to change it back and forth.

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

File: src/api/app/Program.cs [hunk 8]
```diff
@@ -71,7 +76,7 @@
 app.MapControllers();
 
 app.UseExampleModule();
-app.UseBudgetModule();
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/Program.cs:74]: Deleted by mistake, has been corrected.

File: src/api/app/Configuration/SwaggerConfigurationExtensions.cs [hunk 9]
```diff
@@ -14,43 +14,43 @@ public static class SwaggerConfigurationExtensions
 	/// <param name="services">IServiceCollection object.</param>
 	public static void AddSwagger(
 		this IServiceCollection services) => services.AddSwaggerGen(options =>
-	{
```
[the remark this answers is not in the document]
Comment (LukaszWinowicz) [reply src/api/app/Configuration/SwaggerConfigurationExtensions.cs]: Unnecessary whitespace, reverted.

**Review by mateuszmuszynski — APPROVED** [review]

**Pull request merged**